synapse/storage/database.py

```diff
@@ -536,6 +538,31 @@ def new_transaction(
             **kwargs
         """
 
+        # Robustness check: ensure that none of the arguments are generators, since that
+        # will fail if we have to repeat the transaction.
+        # For now, we just log an error, and hope that it works on the first attempt.
+        # TODO: raise an exception.
+        for arg in args:
+            if inspect.isgenerator(arg):
+                logger.error("Programming error: generator passed to new_transaction")
+        for name, val in kwargs.items():
+            if inspect.isgenerator(val):
+                logger.error(
+                    "Programming error: generator passed to new_transaction as argument %s",
+                    name,
+                )
+        # also check variables referenced in func's closure
+        if inspect.isfunction(func):
+            f = cast(types.FunctionType, func)
```

Comment:
mypy doesn't understand that `FunctionType` is a particular type of Callable, hence this business.